main.go

```diff
@@ -29,194 +29,215 @@ import (
-		logger = slog.New(router.Handler())
-		logger = logger.With(slog.String("app", "vault-secret-init"))
+	initLogger(config)
 
-		slog.SetDefault(logger)
+	// Get entrypoint data from arguments
+	binaryPath, binaryArgs, err := args.ExtractEntrypoint(os.Args)
+	if err != nil {
+		slog.Error(fmt.Errorf("failed to extract entrypoint: %w", err).Error())
+		os.Exit(1)
 	}
 
-	provider, err := NewProvider(os.Getenv("PROVIDER"))
+	// Create requested provider and extract relevant secret data
+	// TODO(csatib02): Implement multi-provider support
+	provider, err := NewProvider(config.Provider, config.Daemon)
 	if err != nil {
-		logger.Error(fmt.Errorf("failed to create provider: %w", err).Error())
-
+		slog.Error(fmt.Errorf("failed to create provider: %w", err).Error())
 		os.Exit(1)
 	}
 
-	if len(os.Args) == 1 {
-		logger.Error("no command is given, vault-env can't determine the entrypoint (command), please specify it explicitly or let the webhook query it (see documentation)")
+	envStore := envstore.NewEnvStore()
 
+	providerPaths, err := envStore.GetPathsFor(provider)
+	if err != nil {
+		slog.Error(fmt.Errorf("failed to extract paths: %w", err).Error())
 		os.Exit(1)
 	}
 
-	daemonMode := cast.ToBool(os.Getenv("VAULT_ENV_DAEMON"))
-	delayExec := cast.ToDuration(os.Getenv("VAULT_ENV_DELAY"))
-
-	entrypointCmd := os.Args[1:]
-
-	binary, err := exec.LookPath(entrypointCmd[0])
+	providerSecrets, err := provider.LoadSecrets(context.Background(), providerPaths)
 	if err != nil {
-		logger.Error("binary not found", slog.String("binary", entrypointCmd[0]))
-
+		slog.Error(fmt.Errorf("failed to load secrets: %w", err).Error())
 		os.Exit(1)
 	}
 
-	environ := GetEnvironMap()
-	paths := ExtractPathsFromEnvs(environ)
-
-	ctx := context.Background()
-	secrets, err := provider.LoadSecrets(ctx, paths)
+	secretsEnv, err := envStore.GetProviderSecrets(provider, providerSecrets)
 	if err != nil {
-		logger.Error(fmt.Errorf("failed to load secrets from provider: %w", err).Error())
-
+		slog.Error(fmt.Errorf("failed to convert secrets to envs: %w", err).Error())
 		os.Exit(1)
 	}
-	secretsEnv, err := CreateSecretEnvsFrom(environ, secrets)
-	if err != nil {
-		logger.Error(fmt.Errorf("failed to create environment variables from loaded secrets: %w", err).Error())
 
-		os.Exit(1)
+	// Delay if needed
+	// NOTE(ramizpolic): any specific reason why this is here?
+	if config.Delay > 0 {
+		slog.Info(fmt.Sprintf("sleeping for %s...", config.Delay))
+		time.Sleep(config.Delay)
 	}
 
-	sigs := make(chan os.Signal, 1)
+	slog.Info("spawning process for provided entrypoint command")
 
-	if delayExec > 0 {
-		logger.Info(fmt.Sprintf("sleeping for %s...", delayExec))
-		time.Sleep(delayExec)
+	if !config.Daemon {
+		// When running in non-daemon mode, the process should exit on finish
+		err = syscall.Exec(binaryPath, binaryArgs, secretsEnv)
+		if err != nil {
+			slog.Error(fmt.Errorf("failed to exec process: %w", err).Error())
+			os.Exit(1)
+		}
```

Comment:
missing `return`  statement. If this is not running as deamon, it should not continue further